File: front-end/src/stores/thunks/chats.ts
import { createAsyncThunk } from "@reduxjs/toolkit";
import apiClient from "../../utils/apiClient";
import { AllowedChatListItem, ChatRequestListItem } from "../types";
import { getErrorMessage } from "../../utils";
import { ObjectId } from "../../schemas";

export const fetchAllowedChats = createAsyncThunk('chats/fetchAllowedChats', async (_, thunkAPI) => {
  try {
    const response = await apiClient.post<Array<AllowedChatListItem>>('/chats/allowedChats');
    return response.data;
  } catch (error: unknown) {
    return thunkAPI.rejectWithValue(getErrorMessage(error) || 'Failed to fetch allowed chats');
  }
});

export const fetchChatRequests = createAsyncThunk('chats/fetchChatRequests', async (_, thunkAPI) => {
  try {
    const response = await apiClient.post<Array<ChatRequestListItem>>('/chats/chatRequests');
    return response.data;
  } catch (error: unknown) {
    return thunkAPI.rejectWithValue(getErrorMessage(error) || 'Failed to fetch chat requests');
  }
});

export const approveChatRequest = createAsyncThunk('chats/approve', async (id: ObjectId, { rejectWithValue, dispatch }) => {
  try {
    await apiClient.patch('/chats/approve', { id: id });
  } catch (error: unknown) {
    return rejectWithValue(getErrorMessage(error) || 'Failed to approve chat request');
  } finally {
    dispatch(fetchAllowedChats());
    dispatch(fetchChatRequests());
  }
});

export const rejectChatRequest = createAsyncThunk('chats/reject', async (id: ObjectId, { rejectWithValue, dispatch }) => {
  try {
    await apiClient.patch('/chats/reject', { id: id });
  } catch (error: unknown) {
    return rejectWithValue(getErrorMessage(error) || 'Failed to reject chat request');
  } finally {
    dispatch(fetchChatRequests());
  }
});

export const disallowChat = createAsyncThunk('chats/disallow', async (id: ObjectId, { rejectWithValue, dispatch }) => {
  try {
    await apiClient.patch('/chats/disallow', { id: id });
  } catch (error: unknown) {
    return rejectWithValue(getErrorMessage(error) || 'Failed to disallow chat');
  } finally {
    dispatch(fetchAllowedChats());
    dispatch(fetchChatRequests());
  }
});

File: back-end/app/routes/chats.py
from fastapi import FastAPI, Depends
from fastapi_injector import Injected
from app.models.api import ChatIdRequest
from app.services import ChatsService
from app.utils.jwt_dependencies import jwt_required


def register(app: FastAPI):

    @app.post("/api/chats/allowedChats")
    async def get_chats(
        _ = Depends(jwt_required),
        chats_service = Injected(ChatsService),
    ):
        return await chats_service.get_chats()

    @app.post("/api/chats/chatRequests")
    async def get_chat_requests(
        _ = Depends(jwt_required),
        chats_service = Injected(ChatsService),
    ):
        return await chats_service.get_chat_requests()

    @app.patch("/api/chats/approve")
    async def approve_chat_request(
        body: ChatIdRequest,
        _ = Depends(jwt_required),
        chats_service = Injected(ChatsService),
    ):
        await chats_service.approve_chat_request(body)
        return {"ok": True}

    @app.patch("/api/chats/reject")
    async def reject_chat_request(
        body: ChatIdRequest,
        _ = Depends(jwt_required),
        chats_service = Injected(ChatsService),
    ):
        await chats_service.reject_chat_request(body)
        return {"ok": True}

    @app.patch("/api/chats/disallow")
    async def disallow_chat(
        body: ChatIdRequest,
        _ = Depends(jwt_required),
        chats_service = Injected(ChatsService),
    ):
        await chats_service.disallow_chat(body)
        return {"ok": True}
